Sources/_NIOConcurrency/AsyncSequenceSupport.swift

```diff
@@ -0,0 +1,136 @@
+        
+        @usableFromInline
+        enum State {
+            case hasBuffer(ByteBuffer, Upstream.AsyncIterator)
+            case askForMore(Upstream.AsyncIterator)
+            case finished
+            
+            @inlinable
+            init(buffer: ByteBuffer, upstream: Upstream.AsyncIterator) {
+                if buffer.readableBytes > 0 {
+                    self = .hasBuffer(buffer, upstream)
+                } else {
+                    self = .askForMore(upstream)
+                }
+            }
+        }
+        
+        @inlinable
+        init(_ upstream: Upstream.AsyncIterator) {
+            self.state = .askForMore(upstream)
+        }
+        
+        @inlinable
+        public mutating func next() async throws -> Element? {
+            switch self.state {
+            case .askForMore(var upstream):
+                while true {
+                    switch try await upstream.next() {
+                    case .some(let nextBuffer) where nextBuffer.readableBytes == 0:
+                        // we received an empty buffer. for this reason, let's continue and get the
+                        // next buffer fro, the sequence
+                        continue
+                        
+                    case .some(var nextBuffer):
+                        assert(nextBuffer.readableBytes > 0)
+                        let result = nextBuffer.readInteger(as: UInt8.self)
+                        self.state = .init(buffer: nextBuffer, upstream: upstream)
+                        return result
+                        
+                    case .none:
+                        self.state = .finished
+                        return nil
+                    }
+                }
+                
+            case .hasBuffer(var buffer, let upstream):
+                assert(buffer.readableBytes > 0)
+                let result = buffer.readInteger(as: UInt8.self)
+                self.state = .init(buffer: buffer, upstream: upstream)
+                return result
+                
+            case .finished:
+                return nil
+            }
+        }
+    }
+    
+}
+
+public struct NIOTooManyBytesError: Error {
+    public init() {}
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension AsyncSequence where Element == ByteBuffer {
+    /// Transform an AsyncSequence of ByteBuffers into an AsyncSequence of single bytes.
+    @inlinable
+    public func toBytes() -> NIOByteBufferToUInt8AsyncSequence<Self> {
+        NIOByteBufferToUInt8AsyncSequence(self)
+    }
+    
+    /// Consume an ``Swift/AsyncSequence`` of ``NIO/ByteBuffer``s into a single `ByteBuffer`.
+    ///
+    /// - Parameter maxBytes: The maximum number of bytes that the result ByteBuffer is allowed to have.
+    /// - Returns: A ``NIO/ByteBuffer`` that holds all the bytes of the AsyncSequence
+    @inlinable
+    public func consume(maxBytes: Int) async throws -> ByteBuffer? {
```

Comment:
Hmm, is returning `nil` here a good idea? I'm a bit inclined to want to return an empty buffer.

Reply:
`consume` feels to me like we're actually "reading" the bytes, collect might be more appropriate.

Reply:
I wonder if an additional lower-level variant accepting an `inout ByteBuffer` would also be useful? Callers could for example, reserve capacity or collect from  multiple sequences into the same buffer without an intermediate copy.

Reply:
This seems like it would be more useful as a general "collect bytes from this stream of buffers" API on ByteBuffer itself.

Which means I think there might be benefit to moving it, and loosening the constraints to allow for async sequences of any buffer type (Data, Array<UInt8>, etc).